### DISCUSSION TITLE
Modelling with tide and precipitation

### DISCUSSION THREAD

**Discussion by Damoche01** [pastas/pastas#550] [Q&A] [answered]
Hello all,

Thank you for this innovative idea of pasta. I am modeling the tide's and precipitation's effects on the hydraulic head. The pasta has been helpful, but I noticed my fit is less than average. I have tried Gamma, Polder, and exponential responses. Gamma gave me the best response, but the fit is less than 50%.  So I have the following questions, please:
1. what can you suggest to improve the model?
2. How can I remove the noise from the plot?
3. What is the meaning of recharge_n, recharge_a, recharge_f, and constant_d?
4. How do I explain the step response?
I have attached the adapted pasta code and the figure (Gamma_response) for possible corrections and suggestions.

Additionally, when I tried to input the stress one by one, I got the response attached in the figure (Pasta2)

import pandas as pd
import pastas as ps
import matplotlib.pyplot as plt

oseries = pd.read_csv('SP2_head.csv', index_col='Date', parse_dates=True).squeeze()
rain = pd.read_csv('Rain.csv', index_col='Date', parse_dates=True)
rain['precip']= 0.001*(rain['new_precip'])
waterlevel = pd.read_csv('Tide_pasta.csv', index_col='Date', parse_dates=True)
w=waterlevel['Tide'].mean()
waterlevel_normalize=waterlevel['Tide']-w
waterlevel['waterlevel_normalize']= waterlevel_normalize

ps.plots.series(oseries, [rain['precip'], waterlevel['waterlevel_normalize']],
                figsize=(10, 5), hist=False);

ml = ps.Model(oseries.resample("D").mean().dropna(), name="River")

sm = ps.RechargeModel(rain['precip'], waterlevel['waterlevel_normalize'],
                      rfunc=ps.Gamma(), name="recharge")
ml.add_stressmodel(sm)

ml.solve(tmin="2022-5-6", tmax="2022-12-29")
ml.plots.results(figsize=(12, 8))

I look forward to your response.

![Gamma_response](https://user-images.githubusercontent.com/122163310/218542704-3075834a-55a5-425e-861e-c9946feb0e6b.png)
![Pastas2](https://user-images.githubusercontent.com/122163310/218542707-1d7d3213-c421-432d-8287-b34135534fd6.PNG)

**Comment by HMEUW** [comment, marked as answer]
I advise you to first solve the error about the RechargeModel. The RechargeModel requires both precipitation and evaporation, see https://pastas.readthedocs.io/en/v0.15.0/api/generated/pastas.stressmodels.RechargeModel.html. So add evaporation; or switch to the standard StressModel https://pastas.readthedocs.io/en/v0.15.0/api/generated/pastas.stressmodels.StressModel.html#pastas.stressmodels.StressModel.

**Reply by Damoche01** [reply]
@HMEUW 
Thank you for your suggestion. I plotted the stresses separately, and also combined. I have attached the figures for your reference. Can you kindly look at the figures and the code to see what I can do to improve the model? I will welcome any suggestion.

Here is the code:
import pandas as pd
import pastas as ps
import matplotlib.pyplot as plt

Head = pd.read_csv('SHBL_head.csv', index_col='Date', parse_dates=True).squeeze()
Prec = pd.read_csv('Rain.csv', index_col='Date', parse_dates=True)
Prec['precip']= 0.001*(Prec['new_precip'])
waterlevel = pd.read_csv('Tide_pasta.csv', index_col='Date', parse_dates=True)
w=waterlevel['Tide'].mean()
waterlevel_normalize=waterlevel['Tide']-w
waterlevel['waterlevel_normalize']= waterlevel_normalize

ps.plots.series(Head, [Prec['precip'], waterlevel['waterlevel_normalize']],
                figsize=(10, 5), hist=False);
ml = ps.Model(Head, name="Waterlevels")

sm1 = ps.StressModel(Prec['precip'], rfunc=ps.Gamma(), name="Prec",
                  settings="prec")

sm2 = ps.StressModel(waterlevel['waterlevel_normalize'], rfunc=ps.Gamma(), name="waterlevel",
                  settings="waterlevel")


ml.add_stressmodel([sm1, sm2])
ml.del_noisemodel()

ml.solve(tmin="2022-5-6", tmax="2022-12-30")
ml.plots.results(figsize=(12, 8))

I look forward to suggestions from everyone.

![22222222](https://user-images.githubusercontent.com/122163310/218891443-27e47f87-6048-43df-a197-077dad6dbdae.PNG)
![Combined tides_precip](https://user-images.githubusercontent.com/122163310/218891445-cb1eee3e-a77f-49f9-81a7-61847179f19d.png)
![Precip only](https://user-images.githubusercontent.com/122163310/218891446-bdd6b031-84d3-4561-9e1a-8aa282198c79.png)
![Tides only](https://user-images.githubusercontent.com/122163310/218891449-a4984053-a66c-4039-831f-7199069b91c6.png)

**Reply by martinvonk** [reply]
Hi @Damoche01,

You only fit on a very short amount of time `ml.solve(tmin="2022-5-6", tmax="2022-12-30")`. Is there a reason for this?

Maybe you can post a figure with your stresses here as well? 
`ps.plots.series(Head, stresses=[Prec['precip'], waterlevel['Tide'])`
Maybe accompanied with your .csv files.

**Reply by Damoche01** [reply]
@martinvonk 
Thank you for your response. I have eight months of preliminary data. I limited my time for models (ml.solve(tmin="2022-5-6", tmax="2022-12-30") ) because I do not have verified tidal data that covers the whole length of my well data at the moment. So I set the time to cover the whole length of the available tide for simulation. I have attached the stress figure and the csv files below. Thank you for your help.
I look forward to everyone's contribution and am open to learning.

![Stresses](https://user-images.githubusercontent.com/122163310/219079193-396c7757-dafc-41f6-aa76-acc8d4f0fbf5.png)
[Rain_1.csv](https://github.com/pastas/pastas/files/10744833/Rain_1.csv)
[Tide_pasta.csv](https://github.com/pastas/pastas/files/10744834/Tide_pasta.csv)
[Well_Head.csv](https://github.com/pastas/pastas/files/10744835/Well_Head.csv)

**Reply by martinvonk** [reply]
Hi @Damoche01,

Already by looking at your data, there does not seem to be a relation with the head and the stresses. There is no tide fluctuation in the head so adding the waterlevel won't work. The precipitation level might give an effect but the peaks in head don't seem to (always) correspond to the precipitation events. Nevertheless, I tried some options with your data (shifting the precipitation dates etc.) but I think it won't be possible to create a "good" model unfortunately. Or someone else has to have a good idea. Maybe you have some evaporation / temperature measurements?

I hope this won't discourage you from using Pastas in the future. I still think it is quite a powerful tool and sometime there is no way to create a model that relates the head to the stresses (which is also informative).

**Reply by Damoche01** [reply]
@martinvonk 
Thank you for your contribution. It is really helpful. I do not have temperature data.
Honest question, what will you recommend/suggest to someone that has data that does not have an immediate reflection on the hydraulic head? I will also look forward to other people's responses.
I will never be discouraged from using Pastas. In fact, I have recommended it to some of my connections. Thank you once again. I appreciate it.

**Reply by HMEUW** [reply]
I just had a quick look to your data. My thoughts:

Can you give some background on your setting? The absolute level of the head is approx. 6 m above the water level. What is the distance from your well to the open water?

In the water level I do see some tidal impact, e.g. in the last 2 days. However, it is only a few centimeter.
![image](https://user-images.githubusercontent.com/117160147/219318229-c34b705c-8629-4e8d-8d5c-c0a78319455d.png)

The head has steep peaks. Is the observation correct? Or is there direct infiltration of precipitation in your well?

**Reply by Damoche01** [reply]
@HMEUW Thank you for your effort.
The well is about 2km away from the open sea, and it is in a coastal area. I agree there is tidal impact at some points especially in September which was as a result of 5cm rise in the sea at that time.

**Reply by HMEUW** [reply]
I have not yet used timeseries analysis for tidal impacts, however my first idea is that the impact of the tide is too small on this well. My advise is to check whether the impact is more significant during a storm situation. I might also consider using a higher sampling frequency in the well. I apply a 10 minute interval in my tidal wells.

**Reply by Damoche01** [reply]
@HMEUW Thank you for your suggestion. There is a general big impact on the wells in September. I will also try the 10 minutes you have suggested.

**Reply by Damoche01** [reply]
@martinvonk @HMEUW @dbrakenhoff @mbakker7 
Hi,
Still trying to get my model work. Some of the wells are in good shape while I still need some efforts on the other. It is difficult for me to explain the y-axis of the tide's simulation and its step response. I felt the tide level can never get to 68 m as shown on the tide's simulation, and on the step response, tide's projection to 400 m is worrisome. Can you kindly help me understand it? However, the precipitation's simulation and its step response are understandable and realistic.
I used Gamma as my response function. The tide is in meters, Precipitation in mm/d, and head in m. I have attached the figures for your help. I look forward to your help, discussion and better understanding.
![Well_mo4](https://user-images.githubusercontent.com/122163310/225775514-676bfe58-058d-475c-bdf5-782cce096e9f.png)
![Well_mo4_head](https://user-images.githubusercontent.com/122163310/225775550-ce9d2820-92df-4e0c-a6d4-30eb6c3eea0e.png)

**Reply by HMEUW** [reply]
My suggestion is to create a new model for the tide interaction. The tide effect is small compared to the effect of precipitation. Choose a period of a couple of days without rain.

**Reply by Damoche01** [reply]
Thank you. How can I download the modeled data in pastas?

**Reply by HMEUW** [reply]
What do you mean with 'download'?

**Reply by Damoche01** [reply]
I meant the extraction of the simulated data plotted with the observed head.

**Reply by martinvonk** [reply]
If you want to save the figure:
```python
ax = ml.plot()
fig = ax.get_figure()
fig.savefig(path)
```

If you want to save the simulated values to e.g. a csv:
```python
ml.simulate().to_csv(path)
```

**Reply by Damoche01** [reply]
Thanks a lot

**Reply by Damoche01** [reply]
Hi @HMEUW @martinvonk 
Thanks for your help so far. Is there a way I can save the step response data? If yes, please how. I have tried many things that didn't work.
I have tried:
ml.step_response().to_csv(path)
ml.stepresponse().to_csv(path)

**Reply by martinvonk** [reply]
The function you are looking for is:
https://github.com/pastas/pastas/blob/730251140dab2392b88dc15050f6c2ed006bf4b9/pastas/model.py#L1517-L1524

So if you want the step response:
```python
stepr = ml.get_step_response(name="stressmodelname")
```

Or if you want the block response:
```python
blockr = ml.get_block_response(name="stressmodelname")
```

If you want to get the block / step response on t=0, set the argument `add_0=True`.

These functions return a `pandas.Series` which you can save to a .csv file using the `to_csv()` method if you want.

**Reply by Damoche01** [reply]
Appreciate your help @martinvonk

**Comment by Damoche01** [comment]
@martinvonk @HMEUW @mbakker7 @dbrakenhoff 
I have couple of questions:
1. Is it possible to predict the hydraulic head of 2020 with the 2022 input data using pastas?
2. Each time I tried to pull my step response data, I always get the response below
![1](https://github.com/pastas/pastas/assets/122163310/984fac54-8bdd-4ca5-8a32-d76c5b9125a0)
This is my code:
ml = ps.Model(Head['well1'], name="Head")

sm1 = ps.StressModel(Prec['precipp'], rfunc=ps.Gamma(), name="Prec",
                  settings="prec", up=True)

sm2 = ps.StressModel(Tide['tides'], rfunc=ps.Gamma(), name="Tide",
                  settings="level", up=True)


ml.add_stressmodel([sm1, sm2])

ml.del_noisemodel()

ml.solve()
stepr = ml.get_step_resonse(name="sm1")
stepr1 = ml.get_step_resonse(name="sm2")
ml.plots.results()

 ml.simulate().to_csv('C:/Users/well1simulated_data.csv')
m1.plots.tight_layout()

**Reply by martinvonk** [reply]
1. You can extend the simulation by providing tmin and tmax to ml.simulate(). So in your case: ml.simulate(tmin="2020")

2. `ml.get_step_resonse` should be `ml.get_step_response`

**Reply by Damoche01** [reply]
Thank you @martinvonk 
1. Below is the code, it doesn't extend the simulations to 2020:

Head = pd.read_csv('well1.csv', index_col='Date', parse_dates=True)
Prec = pd.read_csv('precipp.csv', index_col='Date', parse_dates=True)

Tide = pd.read_csv('tides.csv', index_col='Date', parse_dates=True)

ps.plots.series(Head['well1head'], [Prec['precip'], Tide['tide']],
                figsize=(10, 5), hist=False);
ml = ps.Model(Head['well1head'], name="Head")

sm1 = ps.StressModel(Prec['precip'], rfunc=ps.Gamma(), name="Prec",
                  settings="prec", up=True)

sm2 = ps.StressModel(Tide['tide'], rfunc=ps.Gamma(), name="Tide",
                  settings="level", up=True)

ml.simulate(tmin="2020", tmax= "2023")

ml.add_stressmodel([sm1, sm2])
ml.del_noisemodel()
ml.solve()
stepr =  ml.get_step_response(name="sm1")
stepr1 =  ml.get_step_response(name="sm2")
ml.plots.results()
stepr().to_csv('C:/Users/well1_stepresponse.csv')

3. The harvested step response datatype is NoneType, and it not callable.

**Reply by martinvonk** [reply]
1. if you dont have tide and precipitation data till 01-01-2020 you can't extend to 2020
3. you have to provide the stressmodel name to `get_step_response`, not the variable name. so in your case:
`ml.get_step_response("Prec")` and `ml.get_step_response("Tide")`

EDIT: you need stresses data till 2020 to extend simulation, not head

**Reply by Damoche01** [reply]
I appreciate your swift response @martinvonk 
1. This helps. I have rainfall and tide data that extends to 2020 but only have head data of 2022. I will adapt the code above to the stresses. Let me try.
2. I am now able to harvest the step response data.
Thank you for your help and time.

**Reply by martinvonk** [reply]
1. If you have the stresses starting at 2020 that should be fine. You don't need head data for that period if you are extrapolating the head. Can you show what happens if you plot `ml.plot(tmin="2020")`?

**Reply by Damoche01** [reply]
@martinvonk 
1. Do you mean after the input of 2022-2020 stresses data? I am trying to quality check my stresses data of 2021 and 2020. I will post the figure after plotting it.

**Reply by martinvonk** [reply]
1. Yes so you need to have stress data if you want to simulate the head. From checking the .csv data you posted earlier, the tide and precipitation start date is 2022-05-01. Your head data starts on 44687.5 (I assume Excel time -> 2022-05-06). This means you can only let Pastas simulate the head up to 2022-05-01. For dates before that (if you provide tmin="2020"), Pastas will use the `fill_before` time series settings from `ps.rcParams["timeseries"]` supplied to the StressModel under argument `settings`.

**Comment by Damoche01** [comment]
@martinvonk 
![May20_2021-Jan19_2023SP2simulated_data](https://github.com/pastas/pastas/assets/122163310/b329492f-605e-4894-9cb0-50d8446bc662)
![May20_2021-Jan19_2023MO2simulated_data](https://github.com/pastas/pastas/assets/122163310/33c148e5-b2dc-4bdb-b251-b85507d9d14e)

Thank you for your help. Here are the simulations to 2021. Can you kindly critique these simulations and make recommendations? I am hoping to go back in time to 2010 if i get the data.

**Reply by martinvonk** [reply]
Can't you get a potential evaporation (or temperature to [calculate](https://github.com/pyet-org/pyet) pot. evaporation) estimate somewhere? Maybe a global dataset @raoulcollenteur ? That should make your model better because now the seasonal dynamics are covered by the "tide" dynamics.  Where are your observation wells located?

**Reply by Damoche01** [reply]
@martinvonk 
Observation wells are at the east coast of the US.

**Reply by martinvonk** [reply]
Okay so finding a temperature / relative humidity / net or solar radiation / wind speed / evaporation time series from a meteo station should not be that hard than. Highly recommend doing that and adding evaporation to your models.

**Reply by Damoche01** [reply]
Okay. I will add it to the model and see the response. Thanks for your insightful comments.

**Reply by Damoche01** [reply]
Hi @martinvonk @HMEUW @mbakker7 @dbrakenhoff 
I got my model to work, and it is almost ready for publication (see sample in Well1_previous model). However, I got more field data and updated my model, but I have been getting funny outputs even for the field data in pastas (Well_1) and the composite plot of the model and field data (well1_model). When I zoomed in on the model result (well_1 model), the observed data plot was terrible and totally different from the field data (Well1_field observation). I did all I could to check if anything was wrong but couldn't figure it out. Can you kindly look through my script and figures and tell me what is wrong?

This is my script:

from datetime import datetime
import pandas as pd
import pastas as ps
import matplotlib.pyplot as plt

Threshold= 0.7366
rainfall=r'C:/Rainfall_MCAS_1990_Aug19_2023.csv'
tide= r'C:/1hr_New_NAVD88_Tides_1990_Aug19_2023.csv'

date_format = "%m/%d/%Y %H:%M"

Prec = pd.read_csv(rainfall, index_col='Date', parse_dates=True, dayfirst=True)

Tide = pd.read_csv(tide, index_col='Date', parse_dates=True, dayfirst=True)

aj2_head= r'C:/AJ2_hourly.csv'

Head = pd.read_csv(aj2_head, index_col='Date', parse_dates=True, dayfirst=True)
Head['AJ2_Head']=Head['Head']

ps.plots.series(Head['AJ2_Head'], [Prec['Precipitation(mmperday)'], Tide['Tide_height']],
                figsize=(10, 5), hist=False);
ml = ps.Model(Head['AJ2_Head'], name="Head")

sm1 = ps.StressModel(Prec['Precipitation(mmperday)'], rfunc=ps.Gamma(), name="Rain",
                  settings="prec", up=True)

sm2 = ps.StressModel(Tide['Tide_height'], rfunc=ps.FourParam(), name="Tide",
                  settings="level", up=True)

ml.set_parameter("constant_d", initial=0)

ml.simulate(tmin="01-01-1990", tmax= "08-19-2023")

ml.add_stressmodel([sm1, sm2])
ml.del_noisemodel()

ml.solve(solver=ps.LmfitSolve,tmin="01-01-1990", tmax= "08-19-2023")

ml.plots.results(tmin="01-01-1990", tmax= "08-19-2023")

#####Converting the AJ2 model to a dataframe

results3= ml.simulate(tmin="01-01-1990", tmax= "08-19-2023")
if isinstance(results3, pd.Series):
    results3 = results3.to_frame()

results3.reset_index(inplace=True)
aj2_surface_elevation= 5.5

results3['Date'] = pd.to_datetime(results3['index'], format="%Y-%m-%d")
results3.set_index('Date', inplace=True)

impairment= aj2_surface_elevation - Threshold
results3['Flood flag']=(results3['Simulation'])
results3['Flood flag'] =results3['Simulation'].apply(lambda x:1 if x >= impairment else 0)

![Well_1](https://github.com/pastas/pastas/assets/122163310/456a66cc-b5dc-4b49-a288-9af2cd5caa57)
![Well1_field observation](https://github.com/pastas/pastas/assets/122163310/67e13f33-ad45-4c58-8189-ce88f48edf68)
![Well1_model](https://github.com/pastas/pastas/assets/122163310/6e1a909b-51fb-48b2-9c68-d86f4660f111)
![Well1_previous model](https://github.com/pastas/pastas/assets/122163310/e0749aba-0bac-4300-9caa-7d07c9aeeb43)
![well1_zoomed](https://github.com/pastas/pastas/assets/122163310/ca59c4af-7918-45ed-9002-6d834e22785c)

**Reply by martinvonk** [reply]
Looks like there is something wrong with your datetime index of the head series. 
You mention date_format = `"%m/%d/%Y %H:%M"`
However, the keyword argument of `pandas.read_csv(dayfirst=True)` indicates (from the [pandas docstring](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)):
```
dayfirstbool, default False
           DD/MM format dates, international and European format.
```

**Reply by Damoche01** [reply]
@martinvonk 
Thanks for your help. You caught the source of the error and saved me two weeks of hard work checking through what was wrong. I am grateful. Please take a look at the updated model (UpdatedWell) in the attachment.
<img width="535" alt="UpdatedWell" src="https://github.com/pastas/pastas/assets/122163310/68c882b2-766a-4509-bbe6-2baf4c686d4d">
